perses/rjmc/geometry.py

```diff
@@ -2088,47 +2114,188 @@ def _determine_extra_torsions(self, torsion_force, reference_topology, growth_in
+
+        _logger.debug(f"\t\t\trelevant torsions for chirality restraints being added...")
+        #set chirality restraints (adapted from https://github.com/choderalab/perses/blob/protein_mutations_ivy/perses/rjmc/geometry.py)
+
+        #set stereochemistry
+        #the chirality of the atoms is supposed to be pre-specified by NetworkXMolecule
+
+        #render a 3d structure: note that this fucks up the rjmc proposal (since we cannot enumerate the number of possible conformers)
+
+        #add the improper torsions associated with the chiral center
+        coords = self.networkx_structure.mol_oemol.GetCoords()
+        #CIP_perceptions = {0: 'R', 1: 'S'}
+        #iterate over all of the atoms with chiral centers
+        _logger.debug(f"\t\t\t\tnodes: {self.networkx_structure.graph.nodes()}")
+        for _node in self.networkx_structure.graph.nodes(data = True):
+            _logger.debug(f"\t\t\t\tquerying node {_node[0]}")
+            _logger.debug(f"\t\t\t\tnode attributes: {_node[1]}")
+            if _node[1]['oechem_atom'].IsChiral():
+                _logger.debug(f"\t\t\t\tnode is chiral...")
+                assert(_node[1]['oechem_atom']).HasStereoSpecified(), f"atom {_node[1]['oechem_atom']} is chiral, but the chirality is not specified."
+                _stereo = stereo = oechem.OEPerceiveCIPStereo(self.networkx_structure.mol_oemol, _node[1]['oechem_atom'])
+                #_logger.debug(f"\t\t\t\t\tis chiral with CIP: {CIP_perceptions[_stereo]}")
+                #get the neighbors
+                #nbrs_top : list(int) of topology indices
+                #nbrs_oemol : list(int) of oemol indices of neighbor
+                #nbrs : list(OEAtomBase) of the oemol atoms of neighbors
+
+                #get the neighbors of the chiral atom of interest
+                nbrs_top, nbrs_oemol, nbrs = [], [], []
+                for nbr in self.networkx_structure.graph[_node[0]]:
+                    nbrs_top.append(nbr)
+                    nbrs_oemol.append(self.networkx_structure.residue_to_oemol_map[nbr])
+                    nbrs.append(self.networkx_structure.graph.nodes[nbr]['oechem_atom'])
+                _logger.debug(f"\t\t\t\t\tquerying neighbors: {nbrs_top} with data: {[self.networkx_structure.graph.nodes[lst_nbr]['openmm_atom'] for lst_nbr in nbrs_top]}")
+                growth_idx = self._calculate_growth_idx(nbrs_top, growth_indices)
+                _logger.debug(f"\t\t\t\t\tthe growth index of the neighbors is {growth_idx}")
+                if growth_idx > 0:
+
+                    if len(list(self.networkx_structure.graph[_node[0]])) == 4:
+                        _logger.debug(f"\t\t\t\t\tthe number of neighbors is 4; proceeding")
+                        # TODO: handle chiral centers where the valency of the chiral center > 4
+
+                        #specify the atom order for calculating the angle
+
+                        #the order of the improper torsion will be as follows (p1, p2, p3, p4):
```

Comment:
using this improper torsion seems to work for generating proper chiral centers on core and unique new atoms